### DISCUSSION TITLE
Changing `SET NAMES` - why is it discouraged?

### DISCUSSION THREAD

**Discussion by ghnp5** [mydumper/mydumper#1298] [Q&A]
Hey

My application (PHP) runs on every connection:

`SET NAMES utf8mb4 COLLATE utf8mb4_unicode_ci`

(except on earlier years, but it's like this for many years already)

--

When running `myloader`, I'm getting this issue:

> Character set 'utf8mb4_unicode_ci' cannot be used in conjunction with 'binary' in call to regexp_replace.

while it's trying to apply a `CREATE TABLE` that has a `VIRTUAL GENERATED` field that contains `regexp_replace`.

--

If I run the `CREATE TABLE` manually, it runs fine.

However, I've just realized that the problem is that `mydumper` puts this at the top:

`/*!40101 SET NAMES binary*/;`

--

I understand I can change this by including the parameter `--set-names` in the command.

However:

1) The [documentation](https://github.com/mydumper/mydumper/blob/master/docs/mydumper_usage.rst) says:  
> Sets the names, use it at your own risk, default binary

2) In a [comment](https://github.com/mydumper/mydumper/issues/1232#issuecomment-1658354728), @davidducos says:  
> You can use "--set-names utf8mb4" as parameter, which is not the best option because the backup will add the set names to utf8mb4 and it is preferable to use binary, but that is up to you.

--

I'd like to understand why is `binary` preferred and it's so discouraged to change the `--set-names`.

What can go wrong if I set to `utf8mb4`?  
Is it possible to give me an example, please?

Is it just a problem if I have `blob` or `binary` fields?

**EDIT -- Another question is... can I simply replace the schema files so that those have `SET NAMES utf8mb4 COLLATE utf8mb4_unicode_ci` while the data ones have `SET NAMES binary`?**

Thank you very much!

**Comment by davidducos** [comment]
Hi @ghnp5,

I was not able to reproduce your issue. This is my test case:
```
SET NAMES utf8mb4 COLLATE utf8mb4_unicode_ci;
create schema if not exists issue_1298;
use issue_1298;
drop table if exists t1;
create table t1 (id int primary key auto_increment, val varchar(10), val2 varchar(10) as (regexp_replace(val, 'a','b')));
insert into t1(val) values("12345"); insert into t1(val) values("ad"); insert into t1(val) values("bdd"); insert into t1(val) values("bddsdaf");
```
Then I run:
```
root@ubuntu-jammy:~/issue_1298/mydumper# rm -rf data; ./mydumper -o data -T issue_1298.t1 -v 3
root@ubuntu-jammy:~/issue_1298/mydumper# ./myloader -d data/ -v 3 -B issue_1298_2 -o
```
without errors.

Please, provide a test case so I can help you.

**Reply by ghnp5** [reply]
Hi @davidducos 
I see your test case already has `SET NAMES utf8mb4 COLLATE utf8mb4_unicode_ci;`, which is what I said works, and that's what I had to do in my migration script.

However, by default, `mydumper` creates files with:

`/*!40101 SET NAMES binary*/;`

So, when I was running `myloader` against the Percona database, I got the error:

> Character set 'utf8mb4_unicode_ci' cannot be used in conjunction with 'binary' in call to regexp_replace.

The solution was to have my migration script change the `SET NAMES ` in the SQL schema files that contained `regexp_replace`, to `utf8mb4`.

My question is why is changing `SET NAMES` discouraged. What's the worst that could happen.

Many thanks!

**Reply by davidducos** [reply]
Hi @ghnp5 

The schema file for the table is:
```
root@ubuntu-jammy:~/issue_1298/mydumper# cat data/issue_1298.t1-schema.sql
/*!40101 SET NAMES binary*/;
/*!40014 SET FOREIGN_KEY_CHECKS=0*/;
/*!40103 SET TIME_ZONE='+00:00' */;
CREATE TABLE `t1` (
  `id` int NOT NULL AUTO_INCREMENT,
  `val` varchar(10) DEFAULT NULL,
  `val2` varchar(10) GENERATED ALWAYS AS (regexp_replace(`val`,_utf8mb4'a',_utf8mb4'b')) VIRTUAL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=5 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci;
```
myloader is using binary to create the table. I think that the problem is that you are migrating from 5.7 to 8.0 and that 5.7 is not adding the _utf8mb4 conversion.
As I tested manually and I got the error that you mention:
```
mysql> SET NAMES binary;
mysql> create table t1 (id int primary key auto_increment, val varchar(10), val2 varchar(10) as (regexp_replace(val, 'a','b')));
ERROR 3995 (HY000): Character set 'utf8mb4_0900_ai_ci' cannot be used in conjunction with 'binary' in call to regexp_replace.
 ```
I don't know how many funtion do you have with regexp_replace, but I think that will be better to change the GENERATED ALWAYS clause instead of changing the SET NAMES. 
You are not sharing the table structure, so it will be difficult for anyone to determine what is worst that could happen, which at the end is get a mix of charset and collations...

**Reply by ghnp5** [reply]
Thank you @davidducos. I appreciate your help.

Unfortunately it's hard for me to create a minimal example at the moment.

Good thing is that the migration went very well when I made the change (I also had tested it well before I ran in production), so I'm happy that this Thread can be closed.


Maybe the issue was specific to my environment.
The whole MariaDB database was primarily `utf8mb4`, so `binary` is probably very relevant in my environment (unless I'm missing some crucial point).

The Percona database is behaving well and as expected, so I think we're good here!

Thank you again.
